File: apps/api/app/routers/x_to_demo.py
"""X-to-Demo pipeline endpoints."""

from __future__ import annotations

from datetime import datetime
from typing import TYPE_CHECKING

from fastapi import APIRouter, Depends, HTTPException, Response, status

from app.auth import get_current_user
from app.schemas import (
    ResponseEnvelope,
    XToDemoArtifact,
    XToDemoArtifactResponse,
    XToDemoResumeRequest,
    XToDemoRunDetailResponse,
    XToDemoRunRequest,
    XToDemoRunResponse,
    XToDemoUpdateArtifactRequest,
)
from app.services.x_to_demo_pipeline import (
    PhaseKey,
    PipelineArtifact,
    PipelineRunResult,
    XToDemoPipelineService,
    get_x_to_demo_pipeline_service,
)

if TYPE_CHECKING:
    from app.models import User

router = APIRouter(prefix="/x-to-demo", tags=["x-to-demo"])


def get_pipeline_service() -> XToDemoPipelineService:
    """Resolve pipeline service and expose dependency errors as 503s."""
    try:
        return get_x_to_demo_pipeline_service()
    except RuntimeError as exc:
        raise HTTPException(
            status_code=status.HTTP_503_SERVICE_UNAVAILABLE,
            detail=str(exc),
        ) from exc


def _validate_phase_key(phase_key: str) -> PhaseKey:
    if phase_key not in {"feature_spec", "demo_spec", "code_spec"}:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Unsupported phase key. Use one of: feature_spec, demo_spec, code_spec",
        )
    return phase_key  # type: ignore[return-value]


def _artifact_to_schema(artifact: PipelineArtifact) -> XToDemoArtifact:
    return XToDemoArtifact(
        phase_key=artifact.phase_key,
        title=artifact.title,
        markdown=artifact.markdown,
        saved_path=artifact.saved_path,
        json_path=artifact.json_path,
        xml_path=artifact.xml_path,
        json_content=artifact.json_content,
        content_hash=artifact.content_hash,
    )


def _run_result_to_schema(result: PipelineRunResult) -> XToDemoRunResponse:
    return XToDemoRunResponse(
        run_id=result.run_id,
        created_at=result.created_at,
        model=result.model,
        reasoning_effort=result.reasoning_effort,
        artifacts=[_artifact_to_schema(artifact) for artifact in result.artifacts],
        final_code_spec=result.final_code_spec,
        final_code_spec_path=result.final_code_spec_path,
        stop_after_phase=result.stop_after_phase,
        next_phase_key=result.next_phase_key,
        usage_totals=result.usage_totals,
        cost_totals=result.cost_totals,
    )


@router.post(
    "/runs",
    response_model=ResponseEnvelope[XToDemoRunResponse],
    status_code=status.HTTP_201_CREATED,
)
def create_x_to_demo_run(
    request: XToDemoRunRequest,
    current_user: User = Depends(get_current_user),
    pipeline_service: XToDemoPipelineService = Depends(get_pipeline_service),
) -> ResponseEnvelope[XToDemoRunResponse]:
    """Run the X-to-Demo pipeline and return generated artifacts."""
    try:
        result = pipeline_service.run(
            x_input=request.x_input,
            additional_context=request.additional_context,
            feature_name_hint=request.feature_name_hint,
            user_id=current_user.id or 0,
            model=request.model,
            reasoning_effort=request.reasoning_effort,
            stop_after_phase=request.stop_after_phase,
        )
    except ValueError as exc:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=str(exc),
        ) from exc
    except RuntimeError as exc:
        raise HTTPException(
            status_code=status.HTTP_503_SERVICE_UNAVAILABLE,
            detail=str(exc),
        ) from exc
    except Exception as exc:
        raise HTTPException(
            status_code=status.HTTP_502_BAD_GATEWAY,
            detail=f"Pipeline execution failed: {exc}",
        ) from exc

    return ResponseEnvelope(data=_run_result_to_schema(result))


@router.get(
    "/runs/{run_id}",
    response_model=ResponseEnvelope[XToDemoRunDetailResponse],
)
def get_x_to_demo_run(
    run_id: str,
    _current_user: User = Depends(get_current_user),
    pipeline_service: XToDemoPipelineService = Depends(get_pipeline_service),
) -> ResponseEnvelope[XToDemoRunDetailResponse]:
    """Fetch run manifest and current artifact status."""
    try:
        manifest = pipeline_service.get_run_manifest(run_id=run_id)
        result = pipeline_service.get_run_result(run_id=run_id)
    except FileNotFoundError as exc:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(exc)) from exc
    except ValueError as exc:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(exc)) from exc
    except RuntimeError as exc:
        raise HTTPException(
            status_code=status.HTTP_503_SERVICE_UNAVAILABLE, detail=str(exc)
        ) from exc

    phases = manifest.get("phases")
    if not isinstance(phases, list):
        phases = []

    updated_at_raw = manifest.get("updated_at")
    updated_at = result.created_at
    if isinstance(updated_at_raw, str):
        updated_at = datetime.fromisoformat(updated_at_raw.replace("Z", "+00:00"))

    detail = XToDemoRunDetailResponse(
        run_id=result.run_id,
        created_at=result.created_at,
        updated_at=updated_at,
        model=result.model,
        reasoning_effort=result.reasoning_effort,
        stop_after_phase=result.stop_after_phase,
        next_phase_key=result.next_phase_key,
        phases=[
            {
                "phase_key": phase.get("phase_key"),
                "title": phase.get("title"),
                "status": phase.get("status"),
                "input_artifact_ref": phase.get("input_artifact_ref"),
                "output_json_path": phase.get("output_json_path"),
                "output_xml_path": phase.get("output_xml_path"),
                "output_md_path": phase.get("output_md_path"),
                "content_hash": phase.get("content_hash"),
                "error": phase.get("error"),
            }
            for phase in phases
            if isinstance(phase, dict)
        ],
        artifacts=[_artifact_to_schema(artifact) for artifact in result.artifacts],
        usage_totals=result.usage_totals,
        cost_totals=result.cost_totals,
    )
    return ResponseEnvelope(data=detail)


@router.get(
    "/runs/{run_id}/artifacts/{phase_key}",
    response_model=ResponseEnvelope[XToDemoArtifactResponse],
)
def get_x_to_demo_artifact(
    run_id: str,
    phase_key: str,
    _current_user: User = Depends(get_current_user),
    pipeline_service: XToDemoPipelineService = Depends(get_pipeline_service),
) -> ResponseEnvelope[XToDemoArtifactResponse]:
    """Fetch a single phase artifact in markdown + canonical JSON form."""
    resolved_phase = _validate_phase_key(phase_key)
    try:
        artifact = pipeline_service.get_artifact(run_id=run_id, phase_key=resolved_phase)
    except FileNotFoundError as exc:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(exc)) from exc

    return ResponseEnvelope(
        data=XToDemoArtifactResponse(
            run_id=run_id,
            artifact=_artifact_to_schema(artifact),
        )
    )


@router.put(
    "/runs/{run_id}/artifacts/{phase_key}",
    response_model=ResponseEnvelope[XToDemoArtifactResponse],
)
def update_x_to_demo_artifact(
    run_id: str,
    phase_key: str,
    request: XToDemoUpdateArtifactRequest,
    _current_user: User = Depends(get_current_user),
    pipeline_service: XToDemoPipelineService = Depends(get_pipeline_service),
) -> ResponseEnvelope[XToDemoArtifactResponse]:
    """Validate and persist edited artifact content."""
    resolved_phase = _validate_phase_key(phase_key)
    try:
        artifact = pipeline_service.update_artifact(
            run_id=run_id,
            phase_key=resolved_phase,
            markdown=request.markdown,
            json_content=request.json_content,
        )
    except FileNotFoundError as exc:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(exc)) from exc
    except ValueError as exc:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(exc)) from exc

    return ResponseEnvelope(
        data=XToDemoArtifactResponse(
            run_id=run_id,
            artifact=_artifact_to_schema(artifact),
        )
    )


@router.post(
    "/runs/{run_id}/resume",
    response_model=ResponseEnvelope[XToDemoRunResponse],
)
def resume_x_to_demo_run(
    run_id: str,
    request: XToDemoResumeRequest,
    current_user: User = Depends(get_current_user),
    pipeline_service: XToDemoPipelineService = Depends(get_pipeline_service),
) -> ResponseEnvelope[XToDemoRunResponse]:
    """Resume a run from next incomplete (or selected) phase."""
    try:
        result = pipeline_service.resume(
            run_id=run_id,
            user_id=current_user.id or 0,
            from_phase=request.from_phase,
            stop_after_phase=request.stop_after_phase,
            use_edited_artifacts=request.use_edited_artifacts,
        )
    except FileNotFoundError as exc:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(exc)) from exc
    except ValueError as exc:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(exc)) from exc
    except RuntimeError as exc:
        raise HTTPException(
            status_code=status.HTTP_503_SERVICE_UNAVAILABLE, detail=str(exc)
        ) from exc
    except Exception as exc:
        raise HTTPException(
            status_code=status.HTTP_502_BAD_GATEWAY,
            detail=f"Pipeline resume failed: {exc}",
        ) from exc

    return ResponseEnvelope(data=_run_result_to_schema(result))


@router.get("/runs/{run_id}/artifacts/{phase_key}/download")
def download_x_to_demo_artifact(
    run_id: str,
    phase_key: str,
    _current_user: User = Depends(get_current_user),
    pipeline_service: XToDemoPipelineService = Depends(get_pipeline_service),
) -> Response:
    """Download a single artifact markdown file."""
    resolved_phase = _validate_phase_key(phase_key)
    try:
        artifact = pipeline_service.get_artifact(run_id=run_id, phase_key=resolved_phase)
    except FileNotFoundError as exc:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(exc)) from exc

    filename = f"{run_id}-{resolved_phase}.md"
    return Response(
        content=artifact.markdown,
        media_type="text/markdown",
        headers={"Content-Disposition": f'attachment; filename="{filename}"'},
    )


@router.get("/runs/{run_id}/download")
def download_x_to_demo_run(
    run_id: str,
    _current_user: User = Depends(get_current_user),
    pipeline_service: XToDemoPipelineService = Depends(get_pipeline_service),
) -> Response:
    """Download all artifacts + manifest for a run as zip."""
    try:
        archive_bytes = pipeline_service.build_run_download_archive(run_id=run_id)
    except FileNotFoundError as exc:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(exc)) from exc

    return Response(
        content=archive_bytes,
        media_type="application/zip",
        headers={"Content-Disposition": f'attachment; filename="{run_id}-artifacts.zip"'},
    )

File: apps/web/lib/x-to-demo.ts
import { api } from "@/lib/api";
import { AUTH_ACCESS_TOKEN_KEY } from "@/lib/storage-keys";

export const X_TO_DEMO_MODELS = [
  "gpt-5.2",
  "gpt-5.1",
  "gpt-5-mini",
  "gpt-5-nano",
  "gpt-4.1-nano",
] as const;
export type XToDemoModel = (typeof X_TO_DEMO_MODELS)[number];

export const X_TO_DEMO_PHASE_KEYS = ["feature_spec", "demo_spec", "code_spec"] as const;
export type XToDemoPhaseKey = (typeof X_TO_DEMO_PHASE_KEYS)[number];

export const GPT5_REASONING_EFFORTS = ["minimal", "low", "medium", "high"] as const;
export const GPT52_REASONING_EFFORTS = ["none", "low", "medium", "high", "xhigh"] as const;
export type XToDemoReasoningEffort =
  | (typeof GPT5_REASONING_EFFORTS)[number]
  | (typeof GPT52_REASONING_EFFORTS)[number];

export function reasoningEffortsForModel(model: XToDemoModel): readonly XToDemoReasoningEffort[] {
  if (model === "gpt-5.2") {
    return GPT52_REASONING_EFFORTS;
  }
  return GPT5_REASONING_EFFORTS;
}

export interface XToDemoRunRequest {
  x_input: string;
  additional_context?: string;
  feature_name_hint?: string;
  model?: XToDemoModel;
  reasoning_effort?: XToDemoReasoningEffort;
  stop_after_phase?: XToDemoPhaseKey;
}

export interface XToDemoArtifact {
  phase_key: XToDemoPhaseKey;
  title: string;
  markdown: string;
  saved_path: string;
  json_path: string;
  json_content: Record<string, unknown>;
  content_hash: string;
}

export interface XToDemoRunResponse {
  run_id: string;
  created_at: string;
  model: XToDemoModel;
  reasoning_effort: XToDemoReasoningEffort;
  artifacts: XToDemoArtifact[];
  final_code_spec?: string | null;
  final_code_spec_path?: string | null;
  stop_after_phase: XToDemoPhaseKey;
  next_phase_key?: XToDemoPhaseKey | null;
  usage_totals: Record<string, number>;
  cost_totals?: Record<string, number> | null;
}

export interface XToDemoPhaseStatus {
  phase_key: XToDemoPhaseKey;
  title: string;
  status: "pending" | "running" | "completed" | "failed" | "stale";
  input_artifact_ref?: XToDemoPhaseKey | null;
  output_json_path?: string | null;
  output_md_path?: string | null;
  content_hash?: string | null;
  error?: string | null;
}

export interface XToDemoRunDetailResponse {
  run_id: string;
  created_at: string;
  updated_at: string;
  model: XToDemoModel;
  reasoning_effort: XToDemoReasoningEffort;
  stop_after_phase: XToDemoPhaseKey;
  next_phase_key?: XToDemoPhaseKey | null;
  phases: XToDemoPhaseStatus[];
  artifacts: XToDemoArtifact[];
  usage_totals: Record<string, number>;
  cost_totals?: Record<string, number> | null;
}

export interface XToDemoArtifactResponse {
  run_id: string;
  artifact: XToDemoArtifact;
}

export interface XToDemoUpdateArtifactRequest {
  markdown?: string;
  json_content?: Record<string, unknown>;
}

export interface XToDemoResumeRequest {
  from_phase?: XToDemoPhaseKey;
  stop_after_phase?: XToDemoPhaseKey;
  use_edited_artifacts?: boolean;
}

const API_BASE_URL = process.env.NEXT_PUBLIC_API_URL || "http://localhost:8000/api/v1";

async function fetchBinary(path: string): Promise<Blob> {
  const normalizedPath = path.startsWith("/") ? path : `/${path}`;
  const token = typeof window !== "undefined" ? localStorage.getItem(AUTH_ACCESS_TOKEN_KEY) : null;
  const response = await fetch(`${API_BASE_URL}${normalizedPath}`, {
    method: "GET",
    headers: token ? { Authorization: `Bearer ${token}` } : undefined,
  });

  if (!response.ok) {
    throw new Error(`Download failed (${response.status})`);
  }

  return response.blob();
}

export function downloadBlob(blob: Blob, filename: string): void {
  const url = URL.createObjectURL(blob);
  const anchor = document.createElement("a");
  anchor.href = url;
  anchor.download = filename;
  document.body.appendChild(anchor);
  anchor.click();
  anchor.remove();
  URL.revokeObjectURL(url);
}

export async function runXToDemoPipeline(request: XToDemoRunRequest): Promise<XToDemoRunResponse> {
  return api.post<XToDemoRunResponse>("/x-to-demo/runs", request);
}

export async function getXToDemoRun(runId: string): Promise<XToDemoRunDetailResponse> {
  return api.get<XToDemoRunDetailResponse>(`/x-to-demo/runs/${runId}`);
}

export async function getXToDemoArtifact(
  runId: string,
  phaseKey: XToDemoPhaseKey,
): Promise<XToDemoArtifactResponse> {
  return api.get<XToDemoArtifactResponse>(`/x-to-demo/runs/${runId}/artifacts/${phaseKey}`);
}

export async function updateXToDemoArtifact(
  runId: string,
  phaseKey: XToDemoPhaseKey,
  request: XToDemoUpdateArtifactRequest,
): Promise<XToDemoArtifactResponse> {
  return api.put<XToDemoArtifactResponse>(
    `/x-to-demo/runs/${runId}/artifacts/${phaseKey}`,
    request,
  );
}

export async function resumeXToDemoRun(
  runId: string,
  request: XToDemoResumeRequest,
): Promise<XToDemoRunResponse> {
  return api.post<XToDemoRunResponse>(`/x-to-demo/runs/${runId}/resume`, request);
}

export async function downloadXToDemoArtifact(
  runId: string,
  phaseKey: XToDemoPhaseKey,
): Promise<Blob> {
  return fetchBinary(`/x-to-demo/runs/${runId}/artifacts/${phaseKey}/download`);
}

export async function downloadXToDemoRun(runId: string): Promise<Blob> {
  return fetchBinary(`/x-to-demo/runs/${runId}/download`);
}
